mirar/database/transactions/insert.py

```diff
@@ -0,0 +1,52 @@
+"""
+Module to export data to a database table
+"""
+
+import logging
+from typing import Type
+
+import pandas as pd
+from sqlalchemy import Insert, column
+
+from mirar.database.base_table import BaseTable
+from mirar.database.constants import POSTGRES_DUPLICATE_PROTOCOLS
+from mirar.database.engine import get_engine
+
+logger = logging.getLogger(__name__)
+
+
+def _insert_in_table(
+    new_entry: dict,
+    sql_table: Type[BaseTable],
+    duplicate_protocol: str = "fail",
+    returning_keys: list[str] | str = None,
+) -> pd.DataFrame:
+    """
+    Export a list of fields in value dict to a database table
+
+    :param new_entry: dictionary/DataBlock/other dictonary-like object to export
+    :param sql_table: table of DB to export to
+    :param returning_keys: keys to return (default: all)
+    :return:
+    """
+
+    assert duplicate_protocol in POSTGRES_DUPLICATE_PROTOCOLS
+
+    if not isinstance(returning_keys, list):
+        returning_keys = [returning_keys]
```

Comment:
Maybe this is happening elsewhere, but previously every insert statement would return the sequence keys (e.g. uexpid, urawid) to be inserted into image headers/dataframes. Is that still happening? Same q. for update entry.

Reply:
The "sequence keys" are not needed by the code. Some keys are sequence keys, others not, but that is not important. The only thing the code needs to know about are the "primary key" and "unique keys". Whether or not those keys are sequences is irrelevant.

Now, insert returns the primary key by default (uexpid or urawid are the primary keys). So the pipeline works as before.

It does not do that for update though, since in all the pipelines you already do insert first, so you do not need to requery. In a hypothetical world where you did not want to do that, you could select-out the primary key then update.

Reply:
Ok, sounds good. So just trying to understand, if it runs into a unique-conflict while inserting on a table with multiple unique keys, does it decided which row to update by matching the values of all unique keys or just one (the violating one)?

Reply:
It will take every available unique key (i.e ones present in the pydantic model), and update where a match is found to all of them

Reply:
In practise that means it'll match to, for example rawid